packages/firestore/src/core/firestore_client.ts

```diff
@@ -357,6 +367,57 @@ export class FirestoreClient {
     });
   }
 
+  getDocumentFromLocalCache(
+    doc: DocumentReference,
+    observer: Observer<DocumentSnapshot>
+  ): void {
+    this.asyncQueue.schedule(() => {
+      return this.localStore.readDocument(doc._key).then(maybeDoc => {
+        if (maybeDoc instanceof Document) {
+          observer.next(
+            new DocumentSnapshot(doc.firestore, doc._key, maybeDoc, true)
```

Comment:
Per my comments elsewhere, this should return Document instead of creating a DocumentSnapshot.

Reply:
I am a big fan of commenting non-obvious literals in function calls, e.g.:
```
new DocumentSnapshot(doc.firestore, doc._key, maybeDoc, /*fromCache=*/true)
```
We do this in most places (though probably not quite 100%).

Also: It occurs to me that it's possible that the user has an ongoing listener for this document already and so it's guaranteed to be up-to-date, and technically isFromCache should be false in that case. But to make that work, we'd need to route this into the SyncEngine and detect that there's an active View for this document query already, and use the snapshot from that View. We could do this but I'm not sure it's worth the trouble, so I'm fine with keeping this as-is for now. We could consider doing it later.

Reply:
It's actually the EventManager that does the initial snapshot delivery if you relisten:

https://github.com/firebase/firebase-js-sdk/blob/master/packages/firestore/src/core/event_manager.ts#L79